File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
from collections import deque

def solution(book_time):
    answer = 0
    if not book_time: return 0
    def time(t):
        h, m = t.split(":")
        return int(h)*60 + int(m)
    tmp = [(time(start), time(end)) for start, end in book_time]
    tmp = sorted(tmp, key=lambda x: x[0])
    queue, cnt  = deque(), 0
    for t in tmp:
        start, end = t
        while queue and queue[0] <= start: 
            queue.popleft()
            cnt-=1
        cnt +=1
        answer = max(len(queue), answer)
        queue.append(end+10)
        queue = deque(sorted(queue))
    
    return answer+1

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.kt
import java.util.PriorityQueue
import kotlin.math.max

class Solution {
    fun solution(times: Array<Array<String>>): Int {
        var answer = 0
        val rooms = PriorityQueue<Int> { a, b -> a - b }
        val sortedTimes = times.map { (a, b) -> convertTimeToMinutes(a) to convertTimeToMinutes(b) }.sortedBy { it.first }

        for (time in sortedTimes) {
            val (start, end) = time

            if (rooms.isNotEmpty() && rooms.peek() <= start) {
                rooms.poll()
            }

            rooms.add(end + 10)
            answer = max(answer, rooms.size)
        }

        return answer
    }

    private fun convertTimeToMinutes(time: String): Int {
        val (hour, min) = time.split(":").map { it.toInt() }
        return hour * 60 + min
    }
}